instrumentation/opentelemetry-instrumentation-boto3sqs/src/opentelemetry/instrumentation/boto3sqs/__init__.py

```diff
@@ -0,0 +1,426 @@
+        )
+
+    def _wrap_delete_message(self) -> None:
+        def delete_message_wrapper(wrapped, instance, args, kwargs):
+            if receipt_handle := kwargs.get("ReceiptHandle"):
+                Boto3SQSInstrumentor._safe_end_processing_span(receipt_handle)
+            return wrapped(*args, **kwargs)
+
+        wrapt.wrap_function_wrapper(
+            self._sqs_class, "delete_message", delete_message_wrapper
+        )
+
+    def _wrap_delete_message_batch(self) -> None:
+        def delete_message_wrapper_batch(wrapped, instance, args, kwargs):
+            entries = kwargs.get("Entries")
+            for entry in entries:
+                if receipt_handle := entry.get("ReceiptHandle", None):
+                    Boto3SQSInstrumentor._safe_end_processing_span(
+                        receipt_handle
+                    )
+                return wrapped(*args, **kwargs)
+
+        wrapt.wrap_function_wrapper(
+            self._sqs_class,
+            "delete_message_batch",
+            delete_message_wrapper_batch,
+        )
+
+    def _wrap_client_creation(self) -> None:
+        """
+        Since botocore creates classes on the fly using schemas, the SQS class is not necesraily created upon the call
+        of `instrument()`. Therefore we need to wrap the creation of the boto3 client, which triggers the creation of
+        the SQS client.
+        """
+
+        def client_wrapper(wrapped, instance, args, kwargs):
+            retval = wrapped(*args, **kwargs)
+            if not self._did_decorate:
+                self._decorate_sqs()
+            return retval
+
+        wrapt.wrap_function_wrapper(boto3, "client", client_wrapper)
+
+    def _decorate_sqs(self) -> None:
+        """
+        Since botocore creates classes on the fly using schemas, we try to find the class that inherits from the base
+        class and is SQS to wrap.
+        """
+        # We define SQS client as the only client that implements send_message_batch
+        sqs_class = [
+            cls
+            for cls in botocore.client.BaseClient.__subclasses__()
+            if hasattr(cls, "send_message_batch")
+        ]
+        if sqs_class:
+            self._sqs_class = sqs_class[0]
+            self._did_decorate = True
+            self._wrap_send_message()
+            self._wrap_send_message_batch()
+            self._wrap_receive_message()
+            self._wrap_delete_message()
+            self._wrap_delete_message_batch()
+
+    def _un_decorate_sqs(self) -> None:
+        if self._did_decorate:
+            unwrap(self._sqs_class, "send_message")
+            unwrap(self._sqs_class, "send_message_batch")
+            unwrap(self._sqs_class, "receive_message")
+            unwrap(self._sqs_class, "delete_message")
+            unwrap(self._sqs_class, "delete_message_batch")
+            self._did_decorate = False
+
+    def _instrument(self, **kwargs: Dict[str, Any]) -> None:
+        self._did_decorate: bool = False
```

Comment:
why not use the `is_instrumented_by_opentelemetry` function?

Reply:
`is_instrumented_by_opentelemetry` defines a different thing. From the code, its basically a switch that prevents calling `instrument` twice. `self._did_decorate` defines if the functions of `botocore.client.SQS` are wrapped (which is not necessarily true because by calling `instrument` without calling `boto3.client("sqs")` these functions will not be wrapped)